pkg/kubelet/server.go

```diff
@@ -301,6 +308,141 @@ func (s *Server) handleRun(w http.ResponseWriter, req *http.Request) {
 	w.Write(data)
 }
 
+// handleExec handles requests to run a command inside a container.
+func (s *Server) handleExec(w http.ResponseWriter, req *http.Request) {
+	u, err := url.ParseRequestURI(req.RequestURI)
+	if err != nil {
+		s.error(w, err)
+		return
+	}
+	parts := strings.Split(u.Path, "/")
+	var podNamespace, podID, container string
+	var uid types.UID
+	if len(parts) == 5 {
+		podNamespace = parts[2]
+		podID = parts[3]
+		container = parts[4]
+	} else if len(parts) == 6 {
+		podNamespace = parts[2]
+		podID = parts[3]
+		uid = types.UID(parts[4])
+		container = parts[5]
+	} else {
+		http.Error(w, "Unexpected path for command running", http.StatusBadRequest)
+		return
+	}
+	pod, ok := s.host.GetPodByName(podNamespace, podID)
+	if !ok {
+		http.Error(w, "Pod does not exist", http.StatusNotFound)
+		return
+	}
+
+	var (
+		ready                                  sync.WaitGroup
+		inputStream, outputStream, errorStream httpstream.Stream
+	)
+	if req.Header.Get("i") == "1" {
+		ready.Add(1)
+	}
+	if req.Header.Get("o") == "1" {
+		ready.Add(1)
+	}
+	if req.Header.Get("t") != "1" && req.Header.Get("e") == "1" {
+		ready.Add(1)
+	}
+	upgrader := spdy.NewConnectionUpgrader()
+	conn, err := upgrader.UpgradeResponse(w, req, func(stream httpstream.Stream) error {
+		typeString := stream.GetHeader("type")
+		switch typeString {
+		case "input":
+			inputStream = stream
+		case "output":
+			outputStream = stream
+		case "error":
+			errorStream = stream
+		default:
+			return fmt.Errorf("Unknown stream type: %s", typeString)
+		}
+		ready.Done()
+		return nil
+	})
+
+	// TODO we probably should do a timeout here, so it doesn't wait infinitely
+	ready.Wait()
+
+	if inputStream != nil {
+		// close our half of the input stream, since we won't be writing to it
+		inputStream.Close()
+	}
+	command := strings.Split(u.Query().Get("command"), " ")
+	err = s.host.ExecInContainer(GetPodFullName(pod), uid, container, command, inputStream, outputStream, errorStream, req.Header.Get("t") == "1")
+	if err != nil {
+		s.error(w, err)
+		return
+	}
+
+	conn.Close(5 * time.Second)
```

Comment:
It is the time to wait for the streams to be cleaned up before forcefully closing the connection. I think I can get rid of this since I'm now calling Reset() on each stream as part of the connection Close().